pkg/bootstrap/bootstrap.go

```diff
@@ -25,76 +26,100 @@ import (
 
 var releasePattern = regexp.MustCompile("^v[0-9]")
 
-func dataDirFor(dataDir, dataName string) string {
-	return filepath.Join(dataDir, "data", dataName, "bin")
+// binDirForDigest returns the path to dataDir/data/refDigest/bin.
+func binDirForDigest(dataDir string, refDigest string) string {
+	return filepath.Join(dataDir, "data", refDigest, "bin")
 }
 
+// manifestsDir returns the path to dataDir/server/manifests.
 func manifestsDir(dataDir string) string {
 	return filepath.Join(dataDir, "server", "manifests")
 }
 
+// imagesDir returns the path to dataDir/agent/images.
+func imagesDir(dataDir string) string {
+	return filepath.Join(dataDir, "agent", "images")
+}
+
+// symlinkBinDir returns the path to dataDir/bin.
+// This will be symlinked to the current runtime bin dir.
 func symlinkBinDir(dataDir string) string {
 	return filepath.Join(dataDir, "bin")
 }
 
+// dirExists returns true if a directory exists at the given path.
 func dirExists(dir string) bool {
 	if s, err := os.Stat(dir); err == nil && s.IsDir() {
 		return true
 	}
 	return false
 }
 
+// Stage extracts binaries and manifests from the runtime image specified in images into the directory
+// at dataDir. It attempts to load the runtime image from a tarball at dataDir/agent/images,
+// falling back to a remote image pull if the image is not found within a tarball.
+// Extraction is skipped if a bin directory for the specified image already exists.
+// Unique image detection is accomplished by hashing the image name and tag, or the image digest,
+// depending on what the runtime image reference points at.
+// If the bin directory already exists, or content is successfully extracted, the bin directory path is returned.
 func Stage(dataDir string, images images.Images) (string, error) {
 	ref, err := name.ParseReference(images.Runtime)
 	if err != nil {
 		return "", err
 	}
-	// preload image from tarball
-	img, err := preloadBootstrapImage(dataDir, images.Runtime)
+
+	refDigest, err := releaseRefDigest(ref)
 	if err != nil {
 		return "", err
 	}
-	if img == nil {
-		// downloading the image
-		img, err = remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
-		if err != nil {
-			return "", err
-		}
+
+	// Skip all content extraction if the bin dir for this runtime image already exists
+	refBinDir := binDirForDigest(dataDir, refDigest)
+	if dirExists(refBinDir) {
+		logrus.Infof("Runtime image %q bin dir already exists at %q; skipping extract", ref, refBinDir)
+		return refBinDir, nil
 	}
 
-	dataName := releaseName(ref)
-	if dataName != "" {
-		if dir := dataDirFor(dataDir, dataName); dirExists(dir) {
```

Comment:
This path was never taken. releaseName (now releaseRefDigest) always returned an empty string due to bad type assertions, so we never checked for existence and returned early.